### PR TITLE
JIT: one-sided inference from unrelated predicates

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Simplistic version where we just look for identical operands.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Looking at regressions I realized that we sometimes stop looking too early. In earlier versions of `optRedundantBranch` the only partial inferencing we could do was path based, but now it can also be predicate based, so we should keep looking up the tree if we see there's a path. Running new diffs now.

**Comment by AndyAyersMS** [comment]
Also, a fair number of the larger improvements are from removing what appear to be cloned loops.

**Comment by kunalspathak** [comment]
> Also, a fair number of the larger improvements are from removing what appear to be cloned loops.

That's a big win. Just pasting the diff here where we were cloning the loop for redundant condition.

image

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 1]
```diff
@@ -178,15 +178,123 @@ void Compiler::optRelopImpliesRelop(RelopImplicationInfo* rii)
 
     genTreeOps const oper = genTreeOps(funcApp.m_func);
 
+    // Exclude floating point relops.
```
Comment (kunalspathak) [inline src/coreclr/jit/redundantbranchopts.cpp:181]:
```
// Note we could also infer the tree relop's value from other
// dominating relops, for example, (x >= 0) dominating (x > 0).
// That is left as a future enhancement.
```
This comment at the top of function can be deleted now?

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 2]
```diff
@@ -178,15 +178,123 @@ void Compiler::optRelopImpliesRelop(RelopImplicationInfo* rii)
...
+                default:
+                    break;
+
+                case GT_EQ:
+                    rii->canInfer =
+                        (treeOper == GT_GE) || (treeOper == GT_LE) || (treeOper == GT_GT) || (treeOper == GT_LT);
+                    rii->reverseSense      = ((treeOper == GT_GT) || (treeOper == GT_LT));
+                    rii->canInferFromTrue  = true;
+                    rii->canInferFromFalse = false;
+                    rii->vnRelation        = ValueNumStore::VN_RELATION_KIND::VRK_Unrelated;
```
Comment (kunalspathak) [inline src/coreclr/jit/redundantbranchopts.cpp:229]: I am trying to understand the meaning of `VRK_Unrelated` here. What we are saying is although we can infer R* from R, they are not related meaning they cannot be simply swapped or reversed, but they have slightly different operators. Can we have something meaningful enum value for such cases like `VRK_DomInferred` or similar?

Comment (AndyAyersMS) [reply src/coreclr/jit/redundantbranchopts.cpp:229]: Happy to rename it if you think it makes things clearer.

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib PTAL

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 3]
```diff
@@ -370,11 +481,26 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
                         return false;
                     }
 
+                    // Was this an inference from an unrelated relop (GE => GT, say)?
+                    //
+                    const bool domIsUnrelatedRelop = (rii.vnRelation == ValueNumStore::VN_RELATION_KIND::VRK_Unrelated);
```
Comment (kunalspathak) [inline src/coreclr/jit/redundantbranchopts.cpp:486]: likewise here...if you can rename from `domIsUnrelatedRelop` to `domIsInferredRelop`, that would be great.

**Comment by kunalspathak** [comment]
LGTM with few suggestions.

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 4]
```diff
@@ -178,15 +178,123 @@ void Compiler::optRelopImpliesRelop(RelopImplicationInfo* rii)
...
+                    break;
+
+                case GT_LT:
+                    rii->canInfer =
+                        (treeOper == GT_NE) || (treeOper == GT_EQ) || (treeOper == GT_GT) || (treeOper == GT_LE);
+                    rii->reverseSense      = ((treeOper == GT_EQ) || (treeOper == GT_GT));
+                    rii->canInferFromFalse = false;
+                    rii->canInferFromTrue  = true;
+                    rii->vnRelation        = ValueNumStore::VN_RELATION_KIND::VRK_Unrelated;
+                    break;
```
Comment (EgorBo) [inline src/coreclr/jit/redundantbranchopts.cpp:275]: @AndyAyersMS nit: in my attempt to do something very similar I ended up with a sort of a table, like this:
image

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 5]
```diff
@@ -178,15 +178,123 @@ void Compiler::optRelopImpliesRelop(RelopImplicationInfo* rii)
 
     genTreeOps const oper = genTreeOps(funcApp.m_func);
```
Comment (EgorBo) [inline src/coreclr/jit/redundantbranchopts.cpp:179]: I assume it doesn't currently handle unsigned comparisons? (e.g. VNF_GT_UN)

Comment (AndyAyersMS) [reply src/coreclr/jit/redundantbranchopts.cpp:179]: Correct, it does not handle them.

File: src/coreclr/jit/redundantbranchopts.cpp (hunk 4 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/redundantbranchopts.cpp:275]: Yeah, there's probably some refactoring like this that makes sense.

**Comment by AndyAyersMS** [comment]
Ok, revised per feedback. Same diffs as before (locally).

**Review by EgorBo — APPROVED** [review]

**Review by kunalspathak — APPROVED** [review]
Refactoring looks much better now. Thanks!

**Comment by AndyAyersMS** [comment]
Superpmi failure is a timeout.

**Pull request merged**